From 1a013f2a6d170349b4b95e9aae8e629f4223f973 Mon Sep 17 00:00:00 2001
From: Ewan Milne <emilne@redhat.com>
Date: Wed, 20 Aug 2014 18:58:58 -0400
Subject: [scsi] scsi_error: disable eh_deadline if no host_reset_handler is set

Message-id: <1408561176-18133-20-git-send-email-emilne@redhat.com>
Patchwork-id: 88435
O-Subject: [RHEL7.1 PATCH e-stor 19/57] [SCSI] scsi_error: disable eh_deadline if no host_reset_handler is set
Bugzilla: 1132092
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: "Ewan D. Milne" <emilne@redhat.com>

upstream commit ad469a57643b322dc7a3bfc482e265e1e88f735a

Tested by Ewan D. Milne and Barry Donahue with a combination
of manual and automated testing.  No issues seen.

upstream commit ad469a57643b322dc7a3bfc482e265e1e88f735a
Author: Hannes Reinecke <hare@suse.de>
Date:   Fri Jan 17 16:05:38 2014 +0100

    [SCSI] scsi_error: disable eh_deadline if no host_reset_handler is set

    When the host template doesn't declare an eh_host_reset_handler
    the eh_deadline mechanism is pointless and will set the
    device to offline. So disable eh_deadline if no
    eh_host_reset_handler is present.

    Signed-off-by: Hannes Reinecke <hare@suse.de>
    Signed-off-by: James Bottomley <JBottomley@Parallels.com>

Signed-off-by: Ewan D. Milne <emilne@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/hosts.c b/drivers/scsi/hosts.c
index fed9ef1..3cbb57a 100644
--- a/drivers/scsi/hosts.c
+++ b/drivers/scsi/hosts.c
@@ -397,7 +397,8 @@ struct Scsi_Host *scsi_host_alloc(struct scsi_host_template *sht, int privsize)
  shost->use_clustering = sht->use_clustering;
  shost->ordered_tag = sht->ordered_tag;
  shost->no_write_same = sht->no_write_same;
- if (shost_eh_deadline == -1)
+
+ if (shost_eh_deadline == -1 || !sht->eh_host_reset_handler)
   shost->eh_deadline = -1;
  else if ((ulong) shost_eh_deadline * HZ > INT_MAX) {
   shost_printk(KERN_WARNING, shost,
diff --git a/drivers/scsi/scsi_sysfs.c b/drivers/scsi/scsi_sysfs.c
index 196e59a..511f831 100644
--- a/drivers/scsi/scsi_sysfs.c
+++ b/drivers/scsi/scsi_sysfs.c
@@ -300,7 +300,9 @@ store_shost_eh_deadline(struct device *dev, struct device_attribute *attr,
  int ret = -EINVAL;
  unsigned long deadline, flags;
 
- if (shost->transportt && shost->transportt->eh_strategy_handler)
+ if (shost->transportt &&
+     (shost->transportt->eh_strategy_handler ||
+      !shost->hostt->eh_host_reset_handler))
   return ret;
 
  if (!strncmp(buf, "off", strlen("off")))
-- 
1.7.1